integration-test-remote/src/test/java/io/cdap/cdap/app/etl/PubSubTest.java

```diff
@@ -0,0 +1,341 @@
+        return true;
+      } catch (NotFoundException ex) {
+        return false;
+      }
+    }, timeout, unit);
+  }
+
+  private void publishMessages(ProjectTopicName topicName, String... messages) throws Exception {
+    Publisher publisher = Publisher.newBuilder(topicName).setCredentialsProvider(getCredentialProvider()).build();
+    for (String message : messages) {
+      ByteString data = ByteString.copyFromUtf8(message);
+      PubsubMessage pubsubMessage = PubsubMessage.newBuilder().setData(data).build();
+      publisher.publish(pubsubMessage);
+    }
+    publisher.shutdown();
+  }
+
+  private ETLStage createSubscriberStage(String topicName, String subscription) {
+    ArtifactSelectorConfig artifact = new ArtifactSelectorConfig(ArtifactScope.SYSTEM.toString(),
+                                                                 "google-cloud", googleCloudVersion);
+    return new ETLStage(topicName, new ETLPlugin(GOOGLE_SUBSCRIBER_PLUGIN_NAME, StreamingSource.PLUGIN_TYPE,
+                                                 ImmutableMap.of("project", getProjectId(),
+                                                                 "topic", topicName,
+                                                                 "referenceName", "pubsub-input",
+                                                                 "subscription", subscription), artifact));
+  }
+
+  private ETLStage createPublisherStage(String topicName) {
+    ArtifactSelectorConfig artifact = new ArtifactSelectorConfig(ArtifactScope.SYSTEM.toString(),
+                                                                 "google-cloud", googleCloudVersion);
+    return new ETLStage(topicName, new ETLPlugin(GOOGLE_PUBLISHER_PLUGIN_NAME, BatchSink.PLUGIN_TYPE,
+                                                 ImmutableMap.of("project", getProjectId(),
+                                                                 "referenceName", "pubsub-output",
+                                                                 "topic", topicName), artifact));
+  }
+
+  private static class TestMessageReceiver implements MessageReceiver {
+    private List<String> messages = new ArrayList<>();
+    private Gson gson = new Gson();
+
+    @Override
+    public void receiveMessage(PubsubMessage pubsubMessage, AckReplyConsumer ackReplyConsumer) {
+      ackReplyConsumer.ack();
+      String msgJson = pubsubMessage.getData().toStringUtf8();
+      CDAPPubSubMessage msg = gson.fromJson(msgJson, CDAPPubSubMessage.class);
+      String messageString = msg.getMessageString();
+      putMessage(messageString);
+      LOG.info("Retrieved PubSub message: {}", messageString);
+    }
+
+    synchronized List<String> getMessages() {
+      return new ArrayList<>(messages);
+    }
+
+    synchronized int getMessageCount() {
+      return messages.size();
+    }
+
+    private synchronized void putMessage(String message) {
+      messages.add(message);
```

Comment:
you can simply use concurrent version of list

Reply:
Replaced on CopyOnWriteArrayList.